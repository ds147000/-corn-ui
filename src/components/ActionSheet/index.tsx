/* eslint-disable import/export */

import ActionSheet  from './actionSheet'

export { ActionSheetItem } from './item'
export default ActionSheet
