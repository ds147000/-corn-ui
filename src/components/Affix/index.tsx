/* eslint-disable no-unreachable */
/* eslint-disable no-magic-numbers */
// #if _APP === 'weapp'
import Taro, { usePageScroll } from '@tarojs/taro'
// #endif
import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react'
import { View } from '@tarojs/components'
import ClassNames from 'classnames'
import { ViewProps } from '../../types/View'
import { judge } from './utils/judge'

export interface AffixProps extends ViewProps {
  position?: 'top' | 'bottom' | 'right' | 'left'
  onChange?(fixed: boolean): void
}

const Affix: React.FC<AffixProps> = ({ children, position = 'top', onChange, className, ...props }) => {
  const [ fixed, setFixed ] = useState(false)
  const [ boxStyle, setBoxStyle ] = useState<React.CSSProperties>({})
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const ref = useRef<any>()

  const oldFixed = useRef<boolean>(false)

  const _class = useMemo(() => {
    return ClassNames('corn-affix', fixed && `corn-affix-${position}`, className)
  }, [ fixed, position, className ])

  const changeFixed = useCallback((value: boolean): void => {
    setFixed((state) => {
      oldFixed.current = state
      return value
    })
  }, [])

  // #if _APP === 'weapp'
  const system = useRef(Taro.getSystemInfoSync())
  const timer = useRef<NodeJS.Timeout>()

  const checkFlexd = (): void => {
    const node = ref.current

    const rootRect = {
      top: 0,
      left: 0,
      bottom: system.current.windowHeight,
      right: system.current.windowWidth
    } as DOMRectReadOnly

    Taro.createSelectorQuery()
      .select('#' + node.uid)
      .fields({ size: true, rect: true }, (res) => {
        if (!res) return

        setBoxStyle({ height: res?.height, width: res?.width })
        const result = judge(res as unknown as DOMRect, rootRect, position, 0)
        changeFixed(result)
      })
      .exec()
  }

  usePageScroll(() => {
    if (timer.current) clearTimeout(timer.current)
    timer.current = setTimeout(() => checkFlexd(), 10)
  })

  useEffect(() => {
    return (): void => {
      if (timer.current) clearTimeout(timer.current)
    }
  }, [])

  // #else
  useEffect(() => {
    // eslint-disable-next-line no-unreachable
    const el: HTMLDivElement = ref.current
    const rootBounds = {
      x: 0,
      y: 0,
      width: window.innerWidth,
      height: window.innerHeight,
      top: 0,
      left: 0,
      bottom: window.innerHeight,
      right: window.innerWidth
    } as DOMRectReadOnly

    setBoxStyle({ height: el.offsetHeight, width: el.offsetWidth })

    function handleScroll(): void {
      const item = el.getBoundingClientRect()
      const result = judge(item, rootBounds, position, 0)
      changeFixed(result)
    }
    handleScroll()
    window.addEventListener('scroll', handleScroll)
    return (): void => {
      window.removeEventListener('scroll', handleScroll)
    }
  }, [ position, changeFixed ])
  // #endif

  useEffect(() => {
    if (oldFixed.current !== fixed) {
      oldFixed.current = fixed
      onChange?.(fixed)
    }
  }, [ fixed, onChange ])

  return (
    <View style={boxStyle} ref={ref}>
      <View className={_class} {...props} >
        {children}
      </View>
    </View>
  )
}

export default Affix
