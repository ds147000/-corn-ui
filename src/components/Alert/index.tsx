import React from 'react'
import { View } from '@tarojs/components'
import classNames from 'classnames'
import { ViewProps } from '../../types/View'
import Icon from '../Icon'

const Alter: React.FC<ViewProps> = ({ children, className, ...props }) => {
  const _class = classNames(className, 'corn-alert', 'corn-f')
  return (
    <View className={_class} {...props} >
      <Icon name="error" className="icon" />
      {children}
    </View>
  )
}

export default Alter
