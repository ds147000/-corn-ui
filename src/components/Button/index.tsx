import React, { useMemo } from 'react'
import ClassNames from 'classnames'
import { Button as TaroButton } from '@tarojs/components'
import Link, { LinkProps } from '../Link'
import { ITouchEvent } from '../../types'

/** 按钮大小，max: 最大， big: 超大，large：大，middle：中，small：小，mini */
export type ButtonSize = 'max' | 'big' | 'large' | 'middle' | 'small' | 'mini'
export type ButtonType = 'primary' | 'warn' | 'error' | 'link' | 'pop' | 'default' | 'light' | 'stop'

export interface ButtonProps extends LinkProps {
  /** 按钮大小，max: 最大， big: 超大，large：大，middle：中，small：小，mini */
  size?: ButtonSize
  /** 按钮风格类型 */
  type?: ButtonType
  /** 图标 */
  icon?: React.ReactNode
  /** 将按钮宽度调整为其父宽度的选项	 */
  block?: boolean
  /** 镂空风格 */
  ghost?: boolean
  /** 是否禁用 */
  disabled?: boolean
  /** 跳转链接，填写启用A链接行为 */
  to?: string
  /** 表单点击行为 */
  formType?: 'submit' | 'reset'
  /** 开启固定宽度 */
  auto?: boolean
  /** 点击事件 */
  onClick?(e: ITouchEvent): void
}

const Button: React.FC<ButtonProps> = ({
  size = 'middle',
  type = 'primary',
  block,
  ghost,
  disabled,
  to,
  onClick,
  icon,
  children,
  className,
  formType,
  auto,
  ...props
}) => {
  const isIcon = icon !== undefined
  const isChildren = children !== undefined

  const _class = useMemo(() => {
    return ClassNames(
      'corn-btn',
      'corn-if corn-ac corn-jc',
      `corn-btn-${size}`,
      `corn-btn-${type}`,
      {
        'corn-btn-block': block,
        'corn-btn-ghost': ghost,
        'corn-btn-icon': isIcon && !isChildren,
        'corn-btn-child-icon': isIcon && isChildren,
        'corn-btn-disabled': disabled,
        'corn-btn-auto': auto
      },
      className
    )
  }, [ size, type, block, ghost, disabled, className, auto, isIcon, isChildren ])

  const _onClick = (e: ITouchEvent): void => {
    if (disabled) return

    onClick?.(e)
  }



  if (to) {
    return (
      <Link {...props} target="Text" to={to} className={_class} onClick={_onClick} disabled={disabled} type="default" >
        {children}
        {icon}
      </Link>
    )
  }

  return (
    <TaroButton {...props} formType={formType} className={_class} onClick={_onClick} >
      {children}
      {icon}
    </TaroButton>
  )
}

export default Button
