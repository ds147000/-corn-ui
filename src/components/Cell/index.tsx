import React, { useCallback } from 'react'
import ClassNames from 'classnames'
import { View, Text } from '@tarojs/components'
import { ViewProps } from '@tarojs/components/types/View'
import Icon from '../Icon'
import Link, { LinkProps } from '../Link'
import CellList from './list'

export interface CellProps extends ViewProps, LinkProps {
  /** 标题 */
  label?: string | React.ReactNode
  /** 内容 */
  value?: string | React.ReactNode
  /** 内容为空占位提示 */
  placeholder?: string | React.ReactNode
  /** 是否显示箭头 */
  arrow?: boolean
  /** 尾缀内容 */
  suffix?: React.ReactNode | string
  /** 禁用 */
  disabled?: boolean
  /** 跳转链接，填写启用A链接行为。同时会显示箭头 */
  to?: string
}

type CellView =  { List: typeof CellList }



const Cell: React.FC<CellProps> & CellView = ({
  label, value, arrow, suffix,
  disabled, to, children, onClick,
  className, placeholder, ...props
}) => {

  const _onClick = useCallback((event): void => {
    if (disabled) return
    onClick?.(event)
  }, [ onClick, disabled ])

  const _props: ViewProps = {
    className: ClassNames('xrk-cell xrk-f xrk-ac', disabled && 'xrk-cell-disable', className),
    onClick: _onClick,
    ...props
  }

  const showArrow = (arrow || Boolean(to)) && arrow !== false
  const showPlaceholder = Boolean(value) === false && Boolean(children) === false

  const content = (
    <>
      <View className="xrk-cell-label">{label}</View>
      <View className="xrk-cell-value">
        {value}{children}
        {showPlaceholder && <Text className="xrk-cell-placeholder">{placeholder}</Text>}
      </View>
      <View className="xrk-cell-suffix xrk-f xrk-ac">
        {suffix}
        {showArrow && <Icon name="big-right" data-testid="arrow" />}
      </View>
    </>
  )

  if (to)
    return (<Link {..._props} to={to} disabled={disabled} target="View" >{content}</Link>)

  return (<View {..._props} >{content}</View>)
}

Cell.List = CellList

export default Cell
