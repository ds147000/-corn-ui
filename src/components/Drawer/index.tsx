import React, { useCallback, useEffect, useMemo, useState } from 'react'
import classNames from 'classnames'
import { View } from '@tarojs/components'

type AnimateStatus = 'show' | 'hide' | 'over'

export interface DrawerProps {
  /** 是否显示 */
  visible: boolean
  /** 弹出的方向，可以设置左右上下中间, 默认 bottom */
  position?: 'left' | 'right' | 'bottom' | 'top' | 'center'
  /** 是否显示遮罩层 */
  mask?: boolean
  /** 点击蒙层是否允许关闭 */
  maskClosable?: boolean
  /** 关闭回调 */
  onClose?(): void
  /** 完全隐藏回调 */
  onHide?(): void
}

const Drawer: React.FC<DrawerProps> = ({
  visible,
  position = 'bottom',
  mask = true,
  maskClosable = true,
  onClose,
  onHide,
  children
}) => {
  const [ status, setStatus ] = useState<AnimateStatus>('over')
  const _class = useMemo(() => {
    return classNames(
      'xrk-drawer-body',
      status !== 'over' && `xrk-drawer-${position}-${status}`,
      `xrk-drawer-${position}`
    )
  }, [ position, status ])

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const onAnimationEnd = useCallback((event: any) => {
    const animationName: string = event.animationName ||
      event.detail?.animationName ||
      event.nativeEvent?.animationName ||
      event.target.className

    // eslint-disable-next-line no-magic-numbers
    if (animationName.toLocaleUpperCase().indexOf('HIDE') !== -1) setStatus('over')
  }, [])

  const onClickMask = (): void => {
    if (maskClosable) onClose?.()
  }

  useEffect(() => {
    if (visible === true && status !== 'show')
      setStatus('show')

    else if (visible === false && status === 'show')
      setStatus('hide')

  }, [ visible, status ])

  if (visible === false && status === 'over') {
    onHide?.()
    return null
  }

  return (
    <>
      {mask && (
        <View
          className={`xrk-drawer-mask xrk-drawer-mask-${status}`}
          data-testid="mask"
          onClick={onClickMask}
        />
      )}
      <View className={_class} onAnimationEnd={onAnimationEnd} data-testid="body" >
        {children}
      </View>
    </>
  )
}

export default Drawer
