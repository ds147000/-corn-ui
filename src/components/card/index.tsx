/*
 * @Description:
 * @Author: zhoulong.yang
 * @Date: 2021-06-07 12:03:34
 * @LastEditors: zhoulong.yang
 * @LastEditTime: 2021-06-07 15:43:55
 */
import React from 'react'
import { View, Text } from '@tarojs/components'

const Card: React.FC = () => {
  return (
    <View>
      <Text>卡片</Text>
    </View>
  )
}

export default Card
