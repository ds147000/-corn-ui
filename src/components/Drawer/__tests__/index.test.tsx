/* eslint-disable no-magic-numbers */
import React, { useState } from 'react'
import { fireEvent, render, waitFor, createEvent } from '@testing-library/react'
import { View, Text } from '@tarojs/components'
import Drawer, { onTouchMove } from '../index'

describe('Drawer', () => {

  test('hide', async () => {
    const screen = render(
      <Drawer visible={false}>
        <View>
          <Text>Drawer</Text>
        </View>
      </Drawer>
    )

    await waitFor(() => expect(screen.queryByText('Drawer')).toBe(null))
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('show', async () => {
    const screen = render(
      <Drawer visible >
        <View>
          <Text>Drawer</Text>
        </View>
      </Drawer>
    )

    screen.getByText('Drawer')
    await waitFor(() =>  expect(screen.container).toMatchSnapshot())
  })

  test('show of Left', async () => {
    const screen = render(
      <Drawer visible position="left" >
        <View>
          <Text>Drawer</Text>
        </View>
      </Drawer>
    )

    screen.getByText('Drawer')
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('show of Right', async () => {
    const screen = render(
      <Drawer visible position="right" >
        <View>
          <Text>Drawer</Text>
        </View>
      </Drawer>
    )

    screen.getByText('Drawer')
    await waitFor(() => expect(screen.container).toMatchSnapshot())
  })

  test('show of Bottom', () => {
    const screen = render(
      <Drawer visible position="bottom" >
        <View>
          <Text>Drawer</Text>
        </View>
      </Drawer>
    )

    screen.getByText('Drawer')
    expect(screen.container).toMatchSnapshot()
  })

  test('show of Top', () => {
    const screen = render(
      <Drawer visible position="top" >
        <View>
          <Text>Drawer</Text>
        </View>
      </Drawer>
    )

    screen.getByText('Drawer')
    expect(screen.container).toMatchSnapshot()
  })

  test('show of Center', () => {
    const screen = render(
      <Drawer visible position="center" >
        <View>
          <Text>Drawer</Text>
        </View>
      </Drawer>
    )

    screen.getByText('Drawer')
    expect(screen.container).toMatchSnapshot()
  })

  test('mask: false', async () => {
    const screen = render(
      <>
        <Drawer visible position="bottom" mask={false} >
          <View>
            <Text>Drawer</Text>
          </View>
        </Drawer>
      </>
    )

    screen.getByText('Drawer')
    expect(screen.queryByTestId('mask')).toBe(null)
  })

  test('click mask', async () => {
    const onClose = jest.fn()
    const screen = render(
      <>
        <Drawer visible position="bottom" onClose={onClose} >
          <View>
            <Text>Drawer</Text>
          </View>
        </Drawer>
      </>
    )

    screen.getByText('Drawer')
    const mask = screen.getByTestId('mask')
    fireEvent.click(mask)
    await waitFor(() => expect(onClose).toHaveBeenCalledTimes(1))
  })

  test('click mask of maskClosable = false', async () => {
    const onClose = jest.fn()
    const screen = render(
      <>
        <Drawer visible position="bottom" maskClosable={false} onClose={onClose} >
          <View>
            <Text>Drawer</Text>
          </View>
        </Drawer>
      </>
    )

    screen.getByText('Drawer')
    const mask = screen.getByTestId('mask')
    fireEvent.click(mask)
    await waitFor(() => expect(onClose).toHaveBeenCalledTimes(0))
  })

  test('click children', async () => {
    const onClose = jest.fn()
    const screen = render(
      <>
        <Drawer visible position="bottom" onClose={onClose} >
          <View>
            <Text>Drawer</Text>
          </View>
        </Drawer>
      </>
    )

    fireEvent.click(screen.getByText('Drawer'))
    await waitFor(() => expect(onClose).toHaveBeenCalledTimes(0))
  })

  test('change', async () => {
    const onHide = jest.fn()

    const App: React.FC = () => {
      const [ show, setShow ] = useState(false)

      return (
        <>
          <View onClick={(): void => setShow(true)}>显示</View>
          <Drawer visible={show} onClose={(): void => setShow(false)} onHide={onHide} >
            <View>抽屉</View>
          </Drawer>
        </>
      )
    }

    const screen = render(<App />)
    fireEvent.click(screen.getByText('显示'))

    await new Promise((res) => setTimeout(res, 1000))
    fireEvent.animationEnd(screen.getByTestId('body'), { animationName: 'Show' })
    screen.getByText('抽屉')

    fireEvent.click(screen.getByTestId('mask'))
    fireEvent.animationEnd(screen.getByTestId('body'), { animationName: 'Hide' })
    await new Promise((res) => setTimeout(res, 1000))

    await waitFor(() => expect(screen.container).toMatchSnapshot())

    await waitFor(() => expect(onHide).toHaveBeenCalledTimes(1))
  })

  test('onAnimatedDnd', () => {
    const screen = render(<Drawer visible />)
    const evetn = createEvent('animationEnd', screen.getByTestId('body'),
      { animationName: 'Hide' },
      {
        EventType: 'AnimationEvent',
        defaultInit: { bubbles: true, cancelable: false }
      }
    )
    fireEvent.animationEnd(screen.getByTestId('body'), evetn)
  })

  test('touchmove', () => {
    const preventDefault = jest.fn()
    onTouchMove({ preventDefault } as any)

    expect(preventDefault).toHaveBeenCalledTimes(1)
  })

  test('click Children', () => {
    const onClose = jest.fn()
    const screen = render(
      <Drawer visible onClose={onClose} position="center" >
        <div>123</div>
      </Drawer>
    )

    fireEvent.click(screen.getByTestId('body'))
    expect(onClose).toHaveBeenCalledTimes(1)
  })

  test('click Children  bottom', () => {
    const onClose = jest.fn()
    const screen = render(
      <Drawer visible onClose={onClose} >
        <div>123</div>
      </Drawer>
    )

    fireEvent.click(screen.getByTestId('body'))
    expect(onClose).toHaveBeenCalledTimes(0)
  })
})
