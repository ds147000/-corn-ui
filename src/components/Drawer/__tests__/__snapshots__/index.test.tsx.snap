// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer change 1`] = `
<div>
  <div
    class=""
  >
    显示
  </div>
</div>
`;

exports[`Drawer hide 1`] = `<div />`;

exports[`Drawer show 1`] = `
<div>
  <div
    class="corn-drawer-mask corn-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="corn-drawer-body corn-drawer-bottom-show corn-drawer-bottom"
    data-testid="body"
  >
    <div
      class=""
    >
      <span
        class="taro-text"
      >
        Drawer
      </span>
    </div>
  </div>
</div>
`;

exports[`Drawer show of Bottom 1`] = `
<div>
  <div
    class="corn-drawer-mask corn-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="corn-drawer-body corn-drawer-bottom-show corn-drawer-bottom"
    data-testid="body"
  >
    <div
      class=""
    >
      <span
        class="taro-text"
      >
        Drawer
      </span>
    </div>
  </div>
</div>
`;

exports[`Drawer show of Center 1`] = `
<div>
  <div
    class="corn-drawer-mask corn-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="corn-drawer-body corn-drawer-center-show corn-drawer-center"
    data-testid="body"
  >
    <div
      class=""
    >
      <span
        class="taro-text"
      >
        Drawer
      </span>
    </div>
  </div>
</div>
`;

exports[`Drawer show of Left 1`] = `
<div>
  <div
    class="corn-drawer-mask corn-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="corn-drawer-body corn-drawer-left-show corn-drawer-left"
    data-testid="body"
  >
    <div
      class=""
    >
      <span
        class="taro-text"
      >
        Drawer
      </span>
    </div>
  </div>
</div>
`;

exports[`Drawer show of Right 1`] = `
<div>
  <div
    class="corn-drawer-mask corn-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="corn-drawer-body corn-drawer-right-show corn-drawer-right"
    data-testid="body"
  >
    <div
      class=""
    >
      <span
        class="taro-text"
      >
        Drawer
      </span>
    </div>
  </div>
</div>
`;

exports[`Drawer show of Top 1`] = `
<div>
  <div
    class="corn-drawer-mask corn-drawer-mask-show"
    data-testid="mask"
  />
  <div
    class="corn-drawer-body corn-drawer-top-show corn-drawer-top"
    data-testid="body"
  >
    <div
      class=""
    >
      <span
        class="taro-text"
      >
        Drawer
      </span>
    </div>
  </div>
</div>
`;
